src/entities/map/ui/lib/cluster.ts

```diff
@@ -0,0 +1,207 @@
+    };
+  }
+  /**
+   * 특정 마커를 마커 버퍼에 추가 합니다.
+   * 이 때 추가된 마커가 이상값인지 판단하여 이상값 버퍼에 추가 합니다.
+   * ! 이상값으로 판단 된 마커는 클러스터 기준으로 이상값일 뿐
+   * ! 전체 데이터 기준으로 이상값이라고 판단 할 수 없습니다.
+   * ! 오히려 전체 데이터 기준으로 봤을 땐 해당 군집에 어울리는 값일 수 있습니다.
+   * ! 이상값을 판단하는 기준은 Z-Score가 3 이상인 경우로 설정 하였습니다. (3시그마 규칙)
+   */
+  addMarker(marker: T) {
+    const { lat, lng } = marker;
+    // 초기 시행 시에는 따로 Z-Score를 계산하지 않습니다.
+    if (this.std.lat === 0 || this.std.lng === 0) {
+      this.markers.push(marker);
+      return;
+    }
+    const zScore = this.getZScore({ lat, lng });
+    /**
+     * 새로운 군집을 형성 하기 위해 버퍼에 추가합니다.
+     * 이 때 새로운 군집을 형성하는 기준은 이상값이 아닌 마커들로 구성된 군집입니다.
+     */
+    if (zScore.lat > 3 || zScore.lng > 3) {
+      this.outliers.push(marker);
+      return;
+    }
+    this.markers.push(marker);
+  }
+  /**
+   * 마커의 중심점을 재조정 합니다.
+   * 이 때 마커의 중심점이 변경 되었다면 true를 반환 합니다.
+   */
+  revalidateCluster() {
+    // 재조정 전 클러스터의 중심점을 캐싱 합니다.
+    const prevCenter = { ...this.center };
+
+    // 클러스터의 마커들을 이용해 새로운 중심점을 계산합니다.
+    this.center = this.markers.reduce(
+      (center, { lat, lng }) => ({
+        lat: center.lat + lat / this.markers.length,
+        lng: center.lng + lng / this.markers.length,
+      }),
+      { lat: 0, lng: 0 },
+    );
+    // 새로운 중심점을 이용해 통계값을 계산합니다.
+    this.updateStatisticValue();
+    return (
+      prevCenter.lat !== this.center.lat || prevCenter.lng !== this.center.lng
+    );
+  }
+}
+
+export const useKMeansClustering = <T extends LatLng>(
+  NumOfCluster: number,
+  markers?: T[],
+) => {
+  const getCurrentBounds = useGetMapCurrentBounds();
+  const bounds = getCurrentBounds();
+  if (!markers) {
+    return [];
+  }
+
+  // K개의 클러스터를 생성합니다.
+  // TODO 휴리스틱한 방식으로 초기값 뽑기
+  const randomIndexMap: Record<number, boolean> = {};
+  const clusters = Array.from({ length: NumOfCluster }, () => {
+    let randomIndex;
+    do {
+      randomIndex = Math.floor(Math.random() * markers.length);
+    } while (randomIndexMap[randomIndex]);
+    {
+      randomIndexMap[randomIndex] = true;
+    }
+    return new Cluster(markers[randomIndex], bounds);
+  });
+
+  let isChanged = true;
+  while (isChanged) {
+    clusters.forEach((cluster) => cluster.clearMarkers());
+
+    markers.forEach((marker) => {
+      const [, closestClusterIndex] = clusters.reduce(
+        ([minDistance, minClusterIndex], cluster, index) => {
+          const distance = cluster.calculateDistance(marker);
+          return distance < minDistance
+            ? [distance, index]
+            : [minDistance, minClusterIndex];
+        },
+        [Infinity, -1],
+      );
+      // 가장 가까운 클러스터에게 마커를 추가합니다.
+      clusters[closestClusterIndex].addMarker(marker);
+    });
+    // 클러스터의 중심점을 재조정합니다.
+    // 이 때 모든 클러스터의 중심점이 재조정 되지 않았다면 반복문을 종료 합니다.
+    isChanged = clusters.reduce(
+      (isChanged, cluster) => isChanged || cluster.revalidateCluster(),
+      false,
+    );
+  }
+  return clusters;
```

Comment:
reduce로 하신 이유가 있을까요?!
every로도 할 수 있을 것 같아서요


```suggestion
    isChanged = clusters.every(revalidateCluster);
```

Reply:
그르게요 에브리로 하는 편이 더 의미에 맞겠네요 

이거 작업 할 때 리듀스 무새가 되어서 뭐든 리듀스로 했네요 ㅋㅋㅋ,,

Reply:
하나라도 변화가 일어났으면 트루여야 하기에 `some` 으로 변경했습니다. 코드가 훨씬 깔끔해졌네요 !